File: Leetcode/Leetcode-DSA-course/Linked-List/20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> matching = new HashMap<>();
        matching.put('(', ')');
        matching.put('{', '}');
        matching.put('[', ']');
        for (char c : s.toCharArray()) {
            if (matching.containsKey(c))
                stack.push(c);
            else {
                if (stack.empty())
                    return false;
                char previousOpening = stack.pop();
                if (matching.get(previousOpening) != c)
                    return false;
            }
        }
        return stack.empty();
    }
}

File: Leetcode/Leetcode-DSA-course/Linked-List/20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        unordered_map<char, char> matching{
            {'(' , ')'}, {'[' , ']'}, {'{' , '}'}};
        for (char c : s) {
            if (matching.find(c) != matching.end())
                stack.push(c);
            else {
                if (stack.empty())
                    return false;
                char previousOpening = stack.top();
                if (matching[previousOpening] != c)
                    return false;
                stack.pop();
            }
        }
        return stack.empty();
    }
};